### PR TITLE
feat(ts/analyzer): Fix TS error for iterator spread args. Fixes #236.

### LINKED ISSUE #236: Use iterator for spread args
https://github.com/dudykr/stc/blob/798c322abf94b8353feb4f4022185b932069f5e1/crates/stc_ts_type_checker/tests/conformance/es6/spread/iteratorSpreadInCall2.ts#L1-L16

should pass, but `stc` fails with 

```
thread 'conformance::es6::spread::iteratorSpreadInCall2.ts' panicked at '
============================================================
error[TS2345]: DebugContext(
    context: arg is spread
    context:
    lhs = instanceof  symbol[];
    rhs = symbol;
    context: `fail!()` called from assign/mod.rs:1490
    WrongArgType {
        span: Span {
            lo: BytePos(
                234,
            ),
            hi: BytePos(
                252,
            ),
            ctxt: #0,
        },
        inner: AssignFailed {
            span: Span {
                lo: BytePos(
                    234,
                ),
                hi: BytePos(
                    252,
                ),
                ctxt: #0,
            },
            cause: [],
        },
    },
)
  --> $DIR/tests/conformance/es6/spread/iteratorSpreadInCall2.ts:16:8
   |
16 | foo(...new SymbolIterator);
   |        ^^^^^^^^^^^^^^^^^^


    ============================================================
1 unmatched errors out of 1 errors. Got 1 extra errors.
Wanted: [RefError { line: 16, column: 5, code: "TS2556" }]
Unwanted: [(16, "TS2345")]', crates/stc_ts_type_checker/tests/tsc.rs:756:17
```

# Solution (guess)

We should use functions related to iterator, like `get_iterator_element_type` instead of

https://github.com/dudykr/stc/blob/798c322abf94b8353feb4f4022185b932069f5e1/crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs#L2904-L2912

### PR DESCRIPTION
**Pull request by cmrigney**
**Description:**

Returns the correct TS error code for spread parameter.

Please let me know if this is the right approach. I'm new to both Rust and this project. Thanks!

**Related issue:**

 - Closes https://github.com/dudykr/stc/issues/236.

### CODE REVIEW

File: crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs [hunk 1]
```diff
@@ -2932,6 +2932,11 @@ impl Analyzer<'_, '_> {
                     }
 
                     if arg.spread.is_some() {
+                        if let Err(err) = self.get_iterator_element_type(arg.span(), Cow::Borrowed(&arg.ty), false, Default::default()) {
+                            report_err!(Error::SpreadMustBeTupleOrPassedToRest { span: arg.span() });
+                            continue;
+                        }
+
                         match arg.ty.normalize() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs:2940]: What I mentioned in the original issue was this part.

Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs:2940]: The result of `get_iterator_element_type` should be used to return `Ok(())` in the case where element assignment success.

Refer to https://github.com/dudykr/stc/blob/941a9a4981f2d07d84b8e24cc4256576b0e1ac23/crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs#L2943

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

File: crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs (hunk 1 above)
Comment (cmrigney) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs]: Made the changes. Is that what you meant?

File: crates/stc_ts_type_checker/tests/conformance/expressions/functionCalls/callWithSpread2.ts.stats.rust-debug [hunk 2]
```diff
@@ -1,6 +1,6 @@
 Stats {
-    required_error: 5,
-    matched_error: 7,
-    extra_error: 3,
+    required_error: 6,
+    matched_error: 6,
+    extra_error: 4,
```
Comment (kdy1) [inline crates/stc_ts_type_checker/tests/conformance/expressions/functionCalls/callWithSpread2.ts.stats.rust-debug:4]: Can you fix this regression, too?

File: crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs (hunk 1 above)
Comment (kdy1) [reply crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs]: Yes, thank you!

File: crates/stc_ts_type_checker/tests/conformance/expressions/functionCalls/callWithSpread2.ts.stats.rust-debug (hunk 2 above)
Comment (cmrigney) [reply crates/stc_ts_type_checker/tests/conformance/expressions/functionCalls/callWithSpread2.ts.stats.rust-debug]: Fixed! Unfortunately, I had to revert a piece of code that enabled `iteratorSpreadInCall10.ts` to pass. But I think that's a separate issue anyway.

**Review by kdy1 — APPROVED** [review]
Thank you!
Can you rebase and run `./scripts/check.sh` again?

**Comment by kdy1** [comment]
I'll rebase/merge it as I have some time now

**Review by kdy1 — APPROVED** [review]

**Pull request merged**